toolchain/sem_ir/inst.h

```diff
@@ -103,6 +93,10 @@ struct InstLikeTypeInfo<
   }
 };
 
+// A type is InstLike if InstLikeTypeInfo is defined for it.
+template <typename T>
+concept InstLikeType = requires { sizeof(InstLikeTypeInfo<T>); };
```

Comment:
We also use it in the `operator<<` at the bottom of this file, which can't be a member of `Inst`.

It seems like the style guide is motivated in large part by brittleness issues with concepts, which I don't think apply to the toolchain in the same way, given that we don't have any kind of API stability, so I'm not sure to what extent we should follow that guidance. If we want to stick with that, we could replace these concepts (indeed, all the concepts in this PR) with `constexpr bool` templates instead, like they were before, given that we don't intend to allow them to be used as *concept-id*s. What do you think?

Reply:
On the one hand, I agree that we don't quite have the same level of concern currently as the Google style guide is designed for -- that's true for quite a few aspects of it... But I'm hesitant to fully disregard this aspect.

Fundamentally, `sem_ir` really *should* be an API we expect to have users, and I don't think we'd want this to be part of that API directly, and instead be an implementation detail (of that API).

That said, I don't think we need to defend heavily against mis-use, maybe just a detail namespace or 'internal' namespace?

As for switching to `constexpr bool` functions -- I think the risks are basically the same either way. I'd pick between the two based on what makes the definition simpler / easier-to-understand.

Reply:
I meant `constexpr bool` variable templates -- which is what these are today. I don't really see a reason to apply different rules for a `constexpr bool` variable template versus for a `concept`, if we don't care about subsumption (which we don't in this case, because this will be an atomic constraint).

Reply:
I didn't mean to distinguish between variable templates and functions, sorry. And also understand if we have some of these already, mostly trying to figure out how we want to handle things going forward. And so also happy if it shouldn't block this PR but be something we fix forward (to the extent there is any fix).

Anyways, I think the point of the style rule is that when the constraint (whether a constexpr bool variable template or a concept or ...) is likely to be used as a constraint in user-code APIs, it creates overly easily propagated brittleness.

So my feeling is that if this is an issue worth defending against anywhere in the toolchain, it seems likely to be here in what would be a very public API. And I think that would speak to being very cautious about having these in the public API.

If these are important for client code to use as constraints, then maybe they're the rare exception and need to be public concepts. Probably just worth documenting that a bit?

But if they're not used by clients of `sem_ir`, is there harm to putting them in an internal namespace? If so, that seeems like you see an argument to diverge from this aspect of the style guide. I'm open to that I think, but would like to understand it a bit.

Reply:
Actually, reflecting on that a bit -- I think the key difference between what's happening here and the restriction in the style guide is that these concepts are only intended to be applied to exactly the small set of types defined in `typed_insts.h`, not to an open, extensible set of types. Would it make a difference if we guaranteed that somehow?

Reply:
I've moved these into a `Detail` namespace, along with related implementation details in `inst.h`.